File: android/src/main/kotlin/com/example/unified_device_sdk/channels/BleMethodCallHandler.kt
package com.example.unified_device_sdk.channels

import com.example.unified_device_sdk.ble.BleManager
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/** Handles BLE method channel calls and forwards them to [BleManager]. */
class BleMethodCallHandler(
    private val bleManager: BleManager
) : MethodChannel.MethodCallHandler {
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "startScan" -> {
                @Suppress("UNCHECKED_CAST")
                bleManager.startScan(call.arguments as? Map<String, Any?>, result)
            }
            "stopScan" -> bleManager.stopScan(result)
            "connect" -> {
                val deviceId = call.argument<String>("deviceId")
                if (deviceId.isNullOrBlank()) {
                    result.error("invalid_argument", "Missing deviceId", null)
                } else {
                    @Suppress("UNCHECKED_CAST")
                    bleManager.connect(deviceId, call.arguments as? Map<String, Any?>, result)
                }
            }

            "disconnect" -> bleManager.disconnect(result)
            "write" -> {
                val data = call.argument<ByteArray>("data")
                if (data == null || data.isEmpty()) {
                    result.error("invalid_argument", "Missing write payload", null)
                    return
                }
                bleManager.write(data, result)
            }

            else -> result.notImplemented()
        }
    }
}

File: ios/Classes/channels/BleMethodCallHandler.swift
import Flutter
import Foundation

final class BleMethodCallHandler: NSObject, FlutterPlugin {
    private let bleManager: BleManager

    init(bleManager: BleManager) {
        self.bleManager = bleManager
        super.init()
    }

    static func register(with registrar: FlutterPluginRegistrar) {}

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "startScan":
            bleManager.startScan(
                arguments: call.arguments as? [String: Any],
                result: result
            )
        case "stopScan":
            bleManager.stopScan(result: result)
        case "connect":
            guard
                let arguments = call.arguments as? [String: Any],
                let deviceId = arguments["deviceId"] as? String,
                !deviceId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            else {
                result(
                    NativeErrorMapper.flutterError(
                        code: BleConstants.invalidArgumentError,
                        message: "Missing deviceId"
                    )
                )
                return
            }
            bleManager.connect(
                deviceId: deviceId,
                arguments: arguments,
                result: result
            )
        case "disconnect":
            bleManager.disconnect(result: result)
        case "write":
            guard
                let arguments = call.arguments as? [String: Any],
                let typedData = arguments["data"] as? FlutterStandardTypedData
            else {
                result(
                    NativeErrorMapper.flutterError(
                        code: BleConstants.invalidArgumentError,
                        message: "Missing write payload"
                    )
                )
                return
            }
            bleManager.write(data: typedData.data, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}
